utils/kubeclientmetrics/metric.go

```diff
@@ -0,0 +1,228 @@
+package kubeclientmetrics
+
+import (
+	"io/ioutil"
+	"net/http"
+	"path"
+	"regexp"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/client-go/rest"
+
+	"github.com/argoproj/argo-rollouts/utils/unstructured"
+)
+
+type K8sRequestVerb string
+
+const (
+	List    K8sRequestVerb = "List"
+	Get     K8sRequestVerb = "Get"
+	Create  K8sRequestVerb = "Create"
+	Update  K8sRequestVerb = "Update"
+	Patch   K8sRequestVerb = "Patch"
+	Delete  K8sRequestVerb = "Delete"
+	Unknown K8sRequestVerb = "Unknown"
+)
+
+type ResourceInfo struct {
+	Kind       string
+	Namespace  string
+	Name       string
+	Verb       K8sRequestVerb
+	StatusCode int
+}
+
+func (ri ResourceInfo) HasAllFields() bool {
+	return ri.Kind != "" && ri.Namespace != "" && ri.Name != "" && ri.Verb != "" && ri.StatusCode != 0
+}
+
+type metricsRoundTripper struct {
+	roundTripper http.RoundTripper
+	inc          func(ResourceInfo) error
+}
+
+// isGetOrList Uses a path from a request to determine if the request is a GET or LIST. The function tries to find an
+// API version within the path and then calculates how many remaining segments are after the API version. A LIST request
+// has segments for the kind with a namespace and the specific namespace if the kind is a namespaced resource.
+// Meanwhile a GET request has an additional segment for resource name. As a result, a LIST has an odd number of
+// segments while a GET request has an even number of segments.
+func isGetOrList(r *http.Request) K8sRequestVerb {
+	// The following code checks if the path ends with  value of the path is a resource name or kind.
+	// finds the API version in the url and
+	regex, err := regexp.Compile(`v1\w*?(/[a-zA-Z0-9-]*)(/[a-zA-Z0-9-]*)?(/[a-zA-Z0-9-]*)?(/[a-zA-Z0-9-]*)?`)
```

Comment:
Since the expression is static, this should be moved to a package level regex variable that is compiled during package initialization, rather than every time isGetOrList is called. There is a `regexp.MustCompile` which is intended for this.